p2panda-rs/src/operation/traits.rs

```diff
@@ -61,7 +65,30 @@ pub trait AsOperation {
 /// [`StorageProvider`][crate::storage_provider::traits::StorageProvider] implementations should
 /// implement this for a data structure that represents an operation as it is stored in the
 /// database.
-pub trait AsVerifiedOperation {
+pub trait AsVerifiedOperation:
+    Sized + Clone + Send + Sync + Validate + PartialEq + std::fmt::Debug
+{
+    /// Error type for `AsVerifiedOperation`
+    type VerifiedOperationError: 'static + std::error::Error + Send + Sync;
+
+    /// Returns a new `VerifiedOperation` instance.
+    ///
+    /// Use `VerifiedOperation::new_from_entry()` instead if you want to validate that the operation
+    /// was signed by this public key.
+    fn new(
+        public_key: &Author,
+        operation_id: &OperationId,
+        operation: &Operation,
+    ) -> Result<Self, Self::VerifiedOperationError>;
+
+    /// Returns a new `VerifiedOperation` instance constructed from an `EntrySigned`
```

Comment:
This comment somehow suggests a functionality but it is not implemented as it is a trait. I would rephrase it as a suggestion and maybe someday we can even enforce some safety by how we design the traits.

I don't want to open a whole chapter here in this PR on that topic but I wonder how strict we should be around making validation promises and if our current api is reflecting that. It's maybe something to look at again together with the whole validation topic.

(There is a small typo in "oncoded").

Reply:
Yes, very good point.

In this particular case, these methods were only implemented on the struct before, and in this PR I moved them into the trait definition. So maybe the comments snuck across too :grimacing: .... I'll re-write the comment to be more trait-description-like. 

In general, yeh, this would be a very good conversation to have. I particularly get tripped up on where to place constructors. Sometimes they're needed in the trai (like here) but it feels a bit weird for some reason...